tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```
@@ -0,0 +1,594 @@
+    */
+  def apply(decl: TlaOperDecl, inScopeEx: EtcExpr): EtcExpr = {
+    val paramsAndDoms = formalParamsToTypeVars(0, decl.formalParams).map {
+      case (paramName, paramType) =>
+        (paramName, mkConst(BlameRef(decl.body.ID), SetT1(paramType)) )
+    }
+
+    if (decl.name.startsWith("TypeAssumptions")) {
+      // the special form for type annotations over state variables and constants
+      parseTypeAssumptions(decl, inScopeEx)
+    } else {
+      // The type of the lambda is what we want to see as the type of the declaration.
+      // There are two cases: (1) the definition body contains a type annotation, and (2) no type annotation
+      decl.body match {
+        case OperEx(TypingOper.withType, ValEx(TlaStr(typeText)), body) =>
+          // case 1: the definition body contains a type annotation
+          val parsedType =
+            try {
+              type1Parser(typeText)
+            } catch {
+              case e: Type1ParseError =>
+                throw new TypingInputException(s"Parser error in type annotation of ${decl.name}: ${e.msg}")
+            }
+
+          val lambda = mkAbs(ExactRef(body.ID), this (body), paramsAndDoms: _*)
+          val letIn = mkLet(BlameRef(body.ID), decl.name, lambda, inScopeEx)
+          mkTypeDecl(ExactRef(decl.body.ID), decl.name, parsedType, letIn)
+
+        case _ =>
+          // case 2: no type annotation
+          val lambda = mkAbs(ExactRef(decl.body.ID), this (decl.body), paramsAndDoms: _*)
+          mkLet(BlameRef(decl.body.ID), decl.name, lambda, inScopeEx)
+      }
+    }
+  }
+
+  // parse the type annotations inside TypeAssumptions
+  private def parseTypeAssumptions(decl: TlaOperDecl, inScopeEx: EtcExpr): EtcExpr = {
+    decl.body match {
+      case OperEx(TlaBoolOper.and, args @ _*) =>
+        val annotations =
+          args.collect { case OperEx(TypingOper.assumeType, NameEx(name), ValEx(TlaStr(typeText))) =>  (name, typeText) }
```

Won't this raise a run-time match error if any conjunct isn't an instance of  `AssumeType`? iiuc, that would make the subsequent, indirect check for length match unreachable, and it would return an helpful error to the user (and raise noisy warnings during compilation! :)).

How about a wild card match here that raises the `TypingInputException` if the conjunct isn't an `AssumeType`, along with the information that the user needs to identify the specific conjunct?

---

That is the feature of `.collect`. It calls a partial function. If the function is not defined on a sequence element, it would silently ignore the element. It is quite handy for filtering noise by pattern matching. But then you have to check, whether some elements were actually noise.

---

Ah. It seems unhelpful here, IMO, since we lose specificity and then have to do an indirect check in the following line.